src/odemis/gui/comp/overlay/world.py

```diff
@@ -1231,46 +1314,350 @@ def draw(self, ctx, selected, shift=(0, 0), scale=1.0, canvas=None):
                 self._label.font_size = 9
             else:
                 self._label.font_size = 13
                 self._label.weight = cairo.FONT_WEIGHT_BOLD
             ctx.move_to(*b_start)
             ctx.line_to(*b_end)
             ctx.stroke()
+        if font_size:
+            # When exporting a spatial view for print-ready the font size to draw on the fake canvas is given
+            self._label.font_size = font_size
 
-        # Label and ruler have the same colour, highlighted or not
         self._label.colour = self.highlight if selected else self.colour
         self._label.draw(ctx)
-        # self._debug_edges(ctx)
+        self.debug_edges(ctx)
 
 
-MODE_CREATE_RULER = 1
-MODE_SHOW_RULER = 2
+NONE_LABEL_MODE = 0
+EDIT_START_LABEL_LINE = 1
+EDIT_END_LABEL_LINE = 2
+CREATE_LABEL = 3
+EDIT_TEXT_EDGE_MODE = 4
 
-class RulerOverlay(WorldOverlay):
+
+class LabelTool(GenericGadgetLine):
     """
-       Selection overlay that allows for the selection of a ruler in physical coordinates
-       and displays the size of the ruler in meters. It can handle multiple rulers.
+    Represent a "label" in the canvas (as a sub-part of the GadgetOverlay). Used to draw the label
+    and also handle the mouse interaction when dragging/moving the label.
+    """
+
+    def __init__(self, cnvs, p_start_pos, p_end_pos):
+        super(LabelTool, self).__init__(cnvs, p_start_pos, p_end_pos)
+
+        self._label = Label(
+            "",
+            pos=(0, 0),
+            font_size=12,
+            flip=True,
+            align=wx.ALIGN_CENTRE_HORIZONTAL,
+            colour=self.colour,
+            opacity=1.0,
+            deg=None
+        )
+        self.mode = NONE_LABEL_MODE
+        self.motion = False
+        self.text = ''
+        self.last_shiftscale = None  # previous shift & scale of the canvas to know whether it has changed
+
+    def __str__(self):
+        return "Label %g,%g -> %g,%g" % (self.p_start_pos[0], self.p_start_pos[1],
+                                         self.p_end_pos[0], self.p_end_pos[1])
+
+    def start_dragging(self, drag, vpos, tool_mode):
+        """
+        The user can start dragging (creating/editing) the label when the left mouse button is pressed down.
+        Left dragging on the starting point of the label allows us to change the starting point of the line and
+        change the field of interest.
+        Left dragging on the ending point of the label allows us to edit the text of the label.
+        Args:
+            drag: hover mode (HOVER_START, HOVER_END)
+            vpos: the view coordinates of the mouse cursor once left click mouse event is fired
+
+        """
+        super(LabelTool, self).start_dragging(drag, vpos, tool_mode)
+
+        if drag == gui.HOVER_START:
+            self.mode = EDIT_START_LABEL_LINE
+        elif drag == gui.HOVER_END:
+            if tool_mode == MODE_CREATE_LABEL:
+                self.mode = CREATE_LABEL
```

Comment:
This is the only place that you need the `tool_mode`. Moreover the `CREATE_LABEL` mode is there also just for the creation of the label. I think a simpler solution would be to just use `EDIT_END_LABEL_LINE` always, and detect that the text hasn't ever been set. For instance you could set `.text = None` at init, or you could just have a dedicated flag like `._ask_user_for_text = True`. So in `stop_updating_tool()`, you could check if that flag is set, and call `_edit_text()`.